protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManagerImpl.java

```diff
@@ -0,0 +1,132 @@
+                    }
+                });
+    }
+
+    @Override
+    public Future<Void> addEndpoint(final String tenantId, final MqttEndpoint mqttEndpoint) {
+        return getTenantConnections(tenantId)
+                .map(tenantConnections -> {
+                    tenantConnections.addEndpoint(mqttEndpoint);
+                    return null;
+                });
+    }
+
+    @Override
+    public Future<Boolean> closeEndpoint(final String tenantId, final MqttEndpoint mqttEndpoint) {
+
+        return getTenantConnections(tenantId)
+                .map(tenantConnections -> tenantConnections.closeEndpoint(mqttEndpoint))
+                .onSuccess(amqpLinkClosed -> {
+                    if (amqpLinkClosed) {
+                        connectionsPerTenant.remove(tenantId);
+                    }
+                });
+
+    }
+
+    @Override
+    public void closeAllTenants() {
+        connectionsPerTenant.forEach((k, connections) -> connections.closeAllConnections());
+        connectionsPerTenant.clear();
+    }
+
+    @Override
+    public Future<TelemetrySender> getOrCreateTelemetrySender(final String tenantId) {
+        return getAmqpAdapterClientFactory(tenantId).compose(AmqpAdapterClientFactory::getOrCreateTelemetrySender);
+    }
+
+    @Override
+    public Future<EventSender> getOrCreateEventSender(final String tenantId) {
+        return getAmqpAdapterClientFactory(tenantId).compose(AmqpAdapterClientFactory::getOrCreateEventSender);
+    }
+
+    @Override
+    public Future<CommandResponder> getOrCreateCommandResponseSender(final String tenantId) {
+        return getAmqpAdapterClientFactory(tenantId)
+                .compose(AmqpAdapterClientFactory::getOrCreateCommandResponseSender);
+    }
+
+    @Override
+    public Future<MessageConsumer> createDeviceSpecificCommandConsumer(final String tenantId, final String deviceId,
+            final Consumer<Message> messageHandler) {
+
+        return getAmqpAdapterClientFactory(tenantId)
+                .compose(factory -> factory.createDeviceSpecificCommandConsumer(deviceId, messageHandler));
+
+    }
+
+    @Override
+    public Future<MessageConsumer> createCommandConsumer(final String tenantId,
+            final Consumer<Message> messageHandler) {
+
+        return getAmqpAdapterClientFactory(tenantId).compose(factory -> factory.createCommandConsumer(messageHandler));
+
+    }
+
+    private Future<TenantConnections> getTenantConnections(final String tenantId) throws IllegalArgumentException {
+        final TenantConnections tenantConnections = connectionsPerTenant.get(tenantId);
+        if (tenantConnections == null) {
+            return Future.failedFuture("tenant [" + tenantId + "] is not connected");
+        } else {
+            return Future.succeededFuture(tenantConnections);
+        }
+    }
+
+    private Future<AmqpAdapterClientFactory> getAmqpAdapterClientFactory(final String tenantId)
+            throws IllegalStateException, IllegalArgumentException {
```

Comment:
Concerning the `IllegalStateException` case:
Wouldn't it be better to return a failed Future here for that case? I guess, if this exception was thrown here, it would potentially not be handled properly up the call chain.

Reply:
FMPOV the `IllegalStateException` indicates a programming error which should not be "handled", but considered a bug that needs to be fixed, analogous to the usage of `Objects.requireNonNull(obj)` in the Hono code.

Reply:
Hmm, I don't know whether its necessarily a programming error here. If there's a new MQTT connection being initiated for one device, causing `failedTenant.closeAllConnections();` to be invoked because of AMQP adapter connectivity isssue, and if the verticle event loop thread handles an MQTT Publish request from another device around the same time, it isn't immediately obvious to me, that the IllegalStateException can't happen.
Therefore I think, failing the Future and adding a WARN/ERROR log message instead would remove the need to be absolutely sure here, that such an edge case can't occur.

Reply:
I followed your suggestions and additionally refactored some more methods in `TenantConnections` which now return futures as well to be more consistent.